packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -0,0 +1,141 @@
+       * - start: At the start
+       * - end: At the end
+       * @default 'end'
+       */
+      overflowLocation?: 'start' | 'end';
+    }
+  | {
+      children: (visibleCount: number) => React.ReactNode;
+      overflowTag?: undefined;
+      overflowLocation?: undefined;
+    }
+);
+
+/**
+ * Renders fewer children + an `overflowTag` when it starts overflowing. When not overflowing, it renders all children.
+ * This component listens to resize events and updates the rendered content accordingly.
+ *
+ * Two forms of usage:
+ * 1. `children: React.ReactNode[]`: Pass all the children and an `overflowTag` and this component handles when to show
+ * what, depending on whether the component is overflowing.
+ * 2. `children: (visibleCount: number) => React.ReactNode`: For more customization, pass a function to get the
+ * `visibleCount` and then render custom content based on that.
+ *
+ * @example
+ * <OverflowContainer
+ *   as={MyComponent}
+ *   overflowTag={(visibleCount) => (
+ *     <Text>+${tags.length - (visibleCount - 1)} item(s)</Text> // -1 to account for the overflowTag
+ *   )}
+ *   overflowLocation='start'
+ * >
+ *   {items}
+ * </OverflowContainer>
+ *
+ * @example
+ * <OverflowContainer
+ *   as={MyComponent}
+ *   itemsLength={text.length}
+ * >
+ *   {(visibleCount) => {
+ *     // Custom content dependent on visibleCount
+ *     return (
+ *       <>
+ *         itemsLeft(visibleCount)
+ *         overflowButton(visibleCount)
+ *         itemsRight(visibleCount)
+ *       </>
+ *     );
+ *   }
+ * </OverflowContainer>
+ */
+export const OverflowContainer = React.forwardRef((props, ref) => {
+  const {
+    overflowTag,
+    overflowLocation = 'end',
+    items,
+    children,
+    overflowDisabled = false,
+    overflowOrientation,
+    ...rest
+  } = props;
+
+  const [containerRef, visibleCount] = useOverflow(
+    items,
+    overflowDisabled,
+    overflowOrientation,
+  );
+
+  /**
+   * - `visibleCount === children.length` means that we show all children and no overflow tag.
+   * - `visibleCount < children.length` means that we show visibleCount - 1 children and 1 overflow tag.
+   */
+  const visibleItems = React.useMemo(() => {
+    // Consumer wants complete control over what items are rendered.
+    if (typeof children === 'function' || overflowTag == null) {
+      return null;
+    }
+
+    if (visibleCount >= children.length) {
+      return children;
+    }
+
+    if (overflowLocation === 'start') {
+      return (
+        <>
+          {overflowTag(visibleCount)}
+          {children.slice(children.length - (visibleCount - 1))}
+        </>
+      );
+    }
+
+    return (
+      <>
+        {children.slice(0, visibleCount - 1)}
+        {overflowTag(visibleCount)}
+      </>
+    );
+  }, [children, overflowTag, overflowLocation, visibleCount]);
+
+  return (
+    <Box ref={useMergedRefs(ref, containerRef)} {...rest}>
+      {typeof children === 'function' ? children(visibleCount) : visibleItems}
```

Comment:
This render prop branching isn't nice, and kinda defeats the purpose of extracting a component (it's basically a hook with extra steps). What is the main reason for calling `children` as a function?

I'm wondering if it would be better to expose a subcomponent instead.

```jsx
<OverflowContainer>
  <>{/* regular nodes */}</>

  {/* this is rendered only when overflowing */}
  <OverflowNode>…</OverflowNode>
</OverflowContainer>
```

Reply:
Yeah, the new context API makes sense. I think we should also have a hook called something like `useOverflowContext`. It would just be a wrapper over `useSafeContext(OverflowContainerContext)`, with the main goal of allowing to easily destructure the return value without needing to import `useSafeContext` or `OverflowContainerContext`.

My main concern with the context API is the need to extract child components. While it's generally a good idea to have lots of smaller components rather than one large component, it makes the diff too large in some cases (e.g. TablePaginator).